### PR TITLE
[TIR] Fix InjectPTXLDG32 segfaults and skip non-CUDA targets

### LINKED ISSUE #18599: [Bug] Segfault in `tvm.compile (Relax, target=llvm)` during TIR pass `InjectPTXLDG32` / `PTXRewriter::VisitStmt_(BufferStore)` even though target is CPU-only
## Summary

Compiling a Relax IRModule converted from a PyTorch `torch.export` program crashes with a **segmentation fault** inside TVM’s TIR pass pipeline, specifically in:

* `tvm::tir::transform::InjectPTXLDG32(bool)`
* `tvm::tir::PTXRewriter::VisitStmt_(tvm::tir::BufferStoreNode const*)`
* `tvm::tir::BufferStore::BufferStore(...)`

This occurs while invoking `tvm.compile(...)` with:

* `target = tvm.target.Target("llvm")` (CPU-only)
* `tir_pipeline = tir.get_default_tir_pipeline(target)`
* `relax_pipeline = "default"`
* `PassContext.config` includes `"tir.ptx_ldg32": 1` plus several other flags

Even though the target is LLVM CPU, the stack trace indicates a **PTX-specific pass / rewriter** is running and then segfaulting.

This is not a Python exception; it is a hard crash (`Segmentation fault (core dumped)`), so it likely indicates a bug in pass gating / pipeline selection / or an unsafe assumption in the `InjectPTXLDG32` pass when used under this pipeline.

## Environment

From the repro output:

* TVM version: `0.22.0`
* TVM git commit: `9dbf3f22ff6f44962472f9af310fda368ca85ef2`
* LLVM: `17.0.6`
* PyTorch: `2.9.0+cu128`
* Python: `3.10.16` (inferred from stack paths)
* NumPy: `2.2.6`  *(printed as “Python version” in script; see note below)*
* OS: Linux x86_64
* 

## Minimal Repro Script

```python

import random
import numpy as np
import torch
import torch.nn as nn
import tvm
from tvm import tir


def print_env_info():
    print("==== Environment Info ====")
    print("TVM version:", getattr(tvm, "__version__", "unknown"))
    print("TVM git commit:", tvm.support.libinfo()["GIT_COMMIT_HASH"])
    print("TVM LLVM version:", tvm.support.libinfo().get("LLVM_VERSION", "unknown"))

    print("NumPy version:", np.__version__)
    print("PyTorch version:", torch.__version__)
    print("==========================\n")


def set_seed(seed: int = 0):
    random.seed(seed)
    np.random.seed(seed)
    torch.manual_seed(seed)


class Model(nn.Module):
    def __init__(self):
        super().__init__()
        self.net = nn.Sequential(
            nn.ConvTranspose2d(100, 64, 4, 1, 0, bias=False),
            nn.BatchNorm2d(64),
            nn.ReLU(True),
            nn.ConvTranspose2d(64, 3, 4, 2, 1, bias=False),
            nn.Tanh(),
            nn.Conv2d(3, 8, 3, 1, 1, bias=False),
            nn.BatchNorm2d(8),
            nn.LeakyReLU(0.2, inplace=True),
            nn.Conv2d(8, 1, 1, 1, 0, bias=False),
            nn.Sigmoid(),
        )

    def forward(self, x):
        y = self.net(x)
        return y.reshape(-1)


def export_to_relax(mod: nn.Module, x: torch.Tensor) -> tvm.IRModule:
    mod = mod.to("cpu").eval()
    x = x.to("cpu")
    ep = torch.export.export(mod, (x,))
    from tvm.relax.frontend.torch import from_exported_program
    return from_exported_program(ep)


def main():
    print_env_info()
    set_seed(0)

    target = tvm.target.Target("llvm")
    tir_pipeline = tir.get_default_tir_pipeline(target)
    relax_pipeline = "default"

    B = 64
    x = torch.rand(B, 100, 1, 1, dtype=torch.float32)
    model = Model()

    print("[repro] exporting torch -> relax ...")
    ir_mod = export_to_relax(model, x)

    pass_config = {
        "relax.FuseOps.max_depth": 4,
        "relax.backend.use_cuda_graph": 1,
        "tir.disable_storage_rewrite": 1,
        "tir.disable_vectorize": 1,
        "tir.enable_debug": 1,
        "tir.enable_equiv_terms_in_cse_tir": 1,
        "tir.ptx_ldg32": 1,
        "tir.use_async_copy": 1,
    }

    pc_kwargs = {
        "opt_level": 1,
        "disabled_pass": [
            "CanonicalizeBindings",
            "Simplify",
            "VectorizeLoop",
            "RemoveNoOp",
        ],
        "config": pass_config,
    }

    print("[repro] target:", target)
    print("[repro] tir_pipeline: explicit_default")
    print("[repro] PassContext.config keys:", sorted(pass_config.keys()))
    print("[repro] compiling with tvm.compile ...")

    with tvm.transform.PassContext(**pc_kwargs):
        _ = tvm.compile(
            ir_mod,
            target=target,
            relax_pipeline=relax_pipeline,
            tir_pipeline=tir_pipeline,
        )

    print("[repro] compile finished (no crash).")


if __name__ == "__main__":
    main()
```

---

## Actual Behavior

`tvm.compile(...)` crashes with a segfault. The stack trace consistently includes:

* `tvm::tir::BufferStore::BufferStore(...)`
* `tvm::tir::PTXRewriter::VisitStmt_(BufferStoreNode const*)`
* `tvm::tir::transform::InjectPTXLDG32(bool)`

Excerpt:

```text
!!!!!!! Segfault encountered !!!!!!!
...
tvm::tir::BufferStore::BufferStore(...)
tvm::tir::PTXRewriter::VisitStmt_(tvm::tir::BufferStoreNode const*)
...
tvm::tir::transform::PrimFuncPassNode::operator()(...)
...
tvm::tir::transform::InjectPTXLDG32(bool)
Segmentation fault (core dumped)
```

## Triage

* needs-triage
* bug

### LINKED ISSUE #18612: [Bug] Segfault in `tvm.compile` (Relax→TIR, CUDA target) inside `tir::transform::InjectPTXLDG32` / `PTXRewriter::VisitStmt_(BufferStore)` when compiling `torch.export` model returning `(tril, triu)` tuple
## Summary

I am seeing a **hard segfault** (no Python exception) during `tvm.compile(...)` for a CUDA target. The crash consistently occurs inside the TIR pass:

* `tvm::tir::transform::InjectPTXLDG32(bool)`
* `tvm::tir::PTXRewriter::VisitStmt_(tvm::tir::BufferStoreNode const*)`
* `tvm::tir::BufferStore::BufferStore(...)`

The input IRModule is produced by converting a PyTorch `torch.export` program using `tvm.relax.frontend.torch.from_exported_program`. The PyTorch model is intentionally small (`Linear(4,4)`) and returns a **tuple of tensors**: `(torch.tril(x), torch.triu(x))`.

This looks like a bug in the `InjectPTXLDG32` rewrite logic, or an unsafe assumption in the pass leading to a null/invalid `BufferStore` construction.

---

## Environment

From the repro output:

* TVM version: `0.22.0`
* TVM commit: `9dbf3f22ff6f44962472f9af310fda368ca85ef2`
* LLVM: `17.0.6`
* Python: `3.10.16` (from stack paths)
* NumPy: `2.2.6`
* PyTorch: `2.9.0+cu128`
* CUDA GPU: `NVIDIA RTX A6000` (sm_86)

Target string used:

```
cuda -keys=cuda,gpu -arch=sm_86 -max_num_threads=1024 -thread_warp_size=32
```

---

## Minimal Repro Script

```python
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

import numpy as np
import torch
import torch.nn as nn
import tvm
from tvm import tir


def print_env_info():
    print("==== Environment Info ====")
    print("TVM version:", getattr(tvm, "__version__", "unknown"))
    try:
        print("TVM git commit:", tvm.support.libinfo().get("GIT_COMMIT_HASH", "unknown"))
    except Exception:
        print("TVM git commit: unknown")
    try:
        print("TVM LLVM version:", tvm.support.libinfo().get("LLVM_VERSION", "unknown"))
    except Exception:
        print("TVM LLVM version: unknown")
    print("Python (numpy) version:", np.__version__)
    print("PyTorch version:", torch.__version__)
    print("CUDA available (torch):", torch.cuda.is_available())
    if torch.cuda.is_available():
        try:
            print("CUDA device:", torch.cuda.get_device_name(0))
        except Exception:
            pass
    print("==========================\n")


class MyModel(nn.Module):
    def __init__(self):
        super().__init__()
        self.linear = nn.Linear(4, 4)

    def forward(self, x):
        x = self.linear(x)
        return torch.tril(x), torch.triu(x)


def export_to_relax(mod: nn.Module, x: torch.Tensor) -> tvm.IRModule:
    mod = mod.to("cpu").eval()
    x = x.to("cpu")
    ep = torch.export.export(mod, (x,))
    from tvm.relax.frontend.torch import from_exported_program
    return from_exported_program(ep)


def main():
    print_env_info()

    if not torch.cuda.is_available():
        raise RuntimeError("CUDA is required for this repro, but torch.cuda.is_available() is False")

    target_str = "cuda -keys=cuda,gpu -arch=sm_86 -max_num_threads=1024 -thread_warp_size=32"
    target = tvm.target.Target(target_str)

    relax_pipeline = "default"
    tir_pipeline = "default"

    model = MyModel()
    x = torch.zeros((1, 4), dtype=torch.float32)

    print("[repro] exporting torch -> relax ...")
    ir_mod = export_to_relax(model, x)

    disabled_pass = [
        "DeadCodeElimination",
        "CanonicalizeBindings",
        "Simplify",
        "UnrollLoop",
        "VectorizeLoop",
        "StorageRewrite",
        "RemoveNoOp",
        "LoopPartition",
    ]

    pass_config = {
        "relax.FuseOps.max_depth": 2,
        "relax.lift_transform_params.consume_params": 1,
        "tir.disable_storage_rewrite": 1,
        "tir.disable_vectorize": 1,
        "tir.instrument_bound_checkers": 1,
        "tir.merge_static_smem": 1,
        "tir.noalias": 1,
        "tir.ptx_ldg32": 1,
        "tir.use_async_copy": 1,
    }

    pc_kwargs = {
        "opt_level": 3,
        "disabled_pass": disabled_pass,
        "config": pass_config,
    }

    print("[repro] target:", target)
    print("[repro] relax_pipeline:", relax_pipeline)
    print("[repro] tir_pipeline:", tir_pipeline)
    print("[repro] opt_level:", pc_kwargs["opt_level"])
    print("[repro] disabled_pass:", disabled_pass)
    print("[repro] PassContext.config keys:", sorted(pass_config.keys()))
    print("[repro] compiling with tvm.compile ...")

    with tvm.transform.PassContext(**pc_kwargs):
        _ = tvm.compile(
            ir_mod,
            target=target,
            relax_pipeline=relax_pipeline,
            tir_pipeline=tir_pipeline,
        )

    print("[repro] compile finished (no crash).")


if __name__ == "__main__":
    main()
```

---

## Actual Behavior

Segfault during compilation:

```text
!!!!!!! Segfault encountered !!!!!!!
...
tvm::tir::BufferStore::BufferStore(...)
tvm::tir::PTXRewriter::VisitStmt_(tvm::tir::BufferStoreNode const*)
...
tvm::tir::transform::InjectPTXLDG32(bool)
Segmentation fault (core dumped)
```

This is a hard crash (core dumped), not a recoverable error.

---

## Expected Behavior

`tvm.compile(...)` should either:

* successfully compile the module, or
* raise a normal Python exception / diagnostic if some pass config is invalid,

but it should **not segfault**.

### Triage

* needs-triage
* bug

### LINKED ISSUE #18617: [Bug] Segfault in `tvm.compile` on **LLVM (CPU) target** when `tir.ptx_ldg32=1`: unexpectedly runs `tir::transform::InjectPTXLDG32` / `PTXRewriter` and crashes in `BufferStore`
## Summary

`tvm.compile` segfaults when compiling a Relax module (imported from `torch.export`) **even with a pure CPU target** (`llvm -keys=cpu`). The crash occurs inside the PTX-specific pass:

* `tvm::tir::transform::InjectPTXLDG32(bool)`
* `tvm::tir::PTXRewriter::VisitStmt_(BufferStoreNode const*)`
* `tvm::tir::BufferStore::BufferStore(...)`

This is unexpected because the target is **LLVM/CPU**, yet the compilation pipeline still enters a PTX rewriting pass. Removing `tir.ptx_ldg32` from PassContext avoids the crash.

This suggests either:

1. `tir.ptx_ldg32` enables a PTX-only pass without checking whether the target is CUDA/PTX-capable, or
2. `InjectPTXLDG32` lacks a defensive early-exit / target predicate and can crash on non-PTX code paths.

---

## Environment

From the repro output:

* TVM: `0.22.0`
* Commit: `9dbf3f22ff6f44962472f9af310fda368ca85ef2`
* LLVM: `17.0.6`
* Python: `3.10.16` (from stack paths)
* NumPy: `2.2.6`
* PyTorch: `2.9.0+cu128`

Target used in repro:

```
llvm -keys=cpu -mtriple=x86_64-unknown-linux-gnu
```

---

## Reproduction Steps

1. Convert a small PyTorch module to Relax via `torch.export.export` + `tvm.relax.frontend.torch.from_exported_program`.
2. Call `tvm.compile` under a PassContext with `config={"tir.ptx_ldg32": 1}`.
3. Observe segfault during compilation.

---

## Minimal Repro Script

```python
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

import numpy as np
import torch
import torch.nn as nn
import torch.nn.functional as F
import tvm
from tvm import tir


def print_env_info():
    print("==== Environment Info ====")
    print("TVM version:", getattr(tvm, "__version__", "unknown"))
    try:
        li = tvm.support.libinfo()
        print("TVM git commit:", li.get("GIT_COMMIT_HASH", "unknown"))
        print("TVM LLVM version:", li.get("LLVM_VERSION", "unknown"))
    except Exception:
        pass
    print("Python (numpy) version:", np.__version__)
    print("PyTorch version:", torch.__version__)
    print("==========================\n")


class BranchNet(nn.Module):
    def __init__(self, k: int):
        super().__init__()
        self.conv1 = nn.Conv2d(1, 16, k, 1)
        self.conv2 = nn.Conv2d(16, 32, 3, 1)
        self.pool = nn.MaxPool2d(2)

        s1 = 28 - k + 1
        s2 = s1 - 2
        sp = s2 // 2
        self.fc = nn.Linear(32 * sp * sp, 10)

    def forward(self, x):
        x = F.relu(self.conv1(x))
        x = F.relu(self.conv2(x))
        x = self.pool(x)
        x = x.reshape(x.shape[0], -1)
        return self.fc(x)


class M(nn.Module):
    def __init__(self):
        super().__init__()
        self.b1 = BranchNet(3)
        self.b2 = BranchNet(5)
        self.b3 = BranchNet(7)
        self.out = nn.Linear(30, 10)

    def forward(self, x):
        a = self.b1(x)
        b = self.b2(x)
        c = self.b3(x)
        y = self.out(torch.cat([a, b, c], dim=1))
        return F.log_softmax(y, dim=1)


def export_to_relax(mod: nn.Module, x: torch.Tensor) -> tvm.IRModule:
    mod = mod.to("cpu").eval()
    x = x.to("cpu")
    ep = torch.export.export(mod, (x,))
    from tvm.relax.frontend.torch import from_exported_program
    return from_exported_program(ep)


def main():
    print_env_info()

    target = tvm.target.Target("llvm -keys=cpu -mtriple=x86_64-unknown-linux-gnu")
    tir_pipeline = tir.get_default_tir_pipeline(target)  # explicit_default
    relax_pipeline = "default"

    x = torch.rand(1, 1, 28, 28, dtype=torch.float32)
    ir_mod = export_to_relax(M(), x)

    pc = {
        "opt_level": 0,
        "disabled_pass": ["LoopPartition"],
        "config": {
            "tir.ptx_ldg32": 1,
        },
    }

    print("[repro] target:", target)
    print("[repro] tir_pipeline: explicit_default")
    print("[repro] compiling with tvm.compile ...")
    with tvm.transform.PassContext(**pc):
        tvm.compile(ir_mod, target=target, relax_pipeline=relax_pipeline, tir_pipeline=tir_pipeline)


if __name__ == "__main__":
    main()
```

---

## Actual Behavior

Segfault during compilation. Stack trace shows PTX rewrite pass even though target is LLVM:

```text
tvm::tir::BufferStore::BufferStore(...)
tvm::tir::PTXRewriter::VisitStmt_(tvm::tir::BufferStoreNode const*)
...
tvm::tir::transform::InjectPTXLDG32(bool)
Segmentation fault (core dumped)
```

---

## Expected Behavior

On a CPU/LLVM target:

1. Setting `tir.ptx_ldg32=1` should either be ignored (no-op) or rejected with a clear error message, and
2. PTX-specific passes such as `InjectPTXLDG32` should **not run** for non-CUDA targets, and
3. TVM should never segfault; failures should be surfaced as Python exceptions with diagnostics.

### Triage

Please refer to the list of label tags [here](https://github.com/apache/tvm/wiki/Issue-Triage-Labels) to find the relevant tags and add them below in a bullet format (example below).

* needs-triage
* bug

### PR DESCRIPTION
**Pull request by LudovicoYIN**
### Motivation
InjectPTXLDG32 rewrites BufferStore when encountering if_then_else, but it only
initializes temporary buffers when an Allocate node exists. For functions without
Allocate, this leads to uninitialized buffers and a hard segfault during compilation.
In addition, the PTX-only pass can run on CPU/LLVM targets when tir.ptx_ldg32=1,
injecting PTX intrinsics that are invalid for non-CUDA codegen.

This PR ensures temporary buffers are created even when no Allocate exists, and
skips InjectPTXLDG32 on non-CUDA targets, preventing segfaults and invalid PTX
intrinsics on CPU.

### Changes
- Ensure temp buffers are created when the rewrite path is taken without Allocate
- Insert allocations at the function level when needed
- Guard InjectPTXLDG32 so it only runs on CUDA targets
- Add tests for CUDA (insertion) and CPU (skip) behavior

### Testing
test_tir_transform_inject_ptx_ldg32.py

### Fixes
- Fixes #18612
- Fixes #18617
- Fixes #18599

### CODE REVIEW

**Comment by LudovicoYIN** [comment]
@spectrometerHBH @vinx13
This PR fixes segfaults in InjectPTXLDG32 and adds a CUDA-only guard to avoid
invalid PTX intrinsics on CPU targets. All CI is green.
Could you please take a look and approve if it looks good to you? Thanks!
